### ISSUE TITLE
fully promoted soldiers remaining in training sites

### ISSUE THREAD

**Issue by bunnybot** [widelands/widelands#978]
I noticed a bug in 7485. Sometimes, when I reload a game, there is a level 10 soldier in a training site that will not go out on its own. it's quite annoying because it keeps the site stuck. Attached is a saved game where it happens. blue player has 2 training sites, one of them is between the four mountains stretching north to south in the vast green expanse. In that training site, there is a fully promoted soldier inside that will not go out unless kicked manually.

------------------------------------
Imported from Launchpad using lp2gh.

 * date created: 2015-08-29T08:44:58Z
 * owner: lainluigi86
 * assignee: tiborb95
 * the launchpad url was https://bugs.launchpad.net/bugs/1490116

**Comment by bunnybot** [comment]
(by lainluigi86)

**Comment by bunnybot** [comment]
(by w-aaaaa)
duplicate of https://bugs.launchpad.net/widelands/+bug/1428315

**Comment by bunnybot** [comment]
(by tiborb95)
Are you sure this is a problem of game save/load?

I looked at the code, the bottom line is that fully trained soldier is dropped only after succesfull upgrade.  Once an upgrade is not possible (so no upgrade takes place) algorithm is not going to drop such soldier.

It would be convenient if you can say for sure if problem occured during normal (uninterrupted) game or it is problem of save/load.

**Comment by bunnybot** [comment]
(by lainluigi86)
I have no idea. I only noticed the odd behavior after reloading saved games. It doesn't happen very often, so it is difficult to replicate and identify the moment it happpens. Plus, I often play with the training site open, and I never noticed it while I was looking. That's why I suppose it is related to loading a game.

**Comment by bunnybot** [comment]
(by gunchleoc)
I think there should be some safeguard code in any case that the site periodically makes sure that there are no fully trained soldiers hanging around by accident. If this is a saveload bug, such a check will mask that bug, but it would also make the error go away.

**Comment by bunnybot** [comment]
(by tiborb95)
@GunChleoc

I am working on exactly this type of fix. It is good solution, only it does not fix the cause itself. In fact so far I failed to find an actual bug in code...

**Comment by bunnybot** [comment]
(by tiborb95)
fixed in rev. 7516

**Comment by bunnybot** [comment]
(by gunchleoc)
Fixed in build19-rc1.
